### PR TITLE
Add Battery Topic

### PR DESCRIPTION
**Pull request by yukiisbored**
A Battery module is added to make Monika to be able to complain to the
player about the low battery life if the computer have a battery present

Closes https://github.com/Backdash/MonikaModDev/issues/596

TODOs before merging:
- [x] Do tests on Windows
  - [X] Test on a Windows Laptop
  - [x] Test on a Windows Desktop or Laptop with the battery unplugged
- [ ] Do tests on macOS
- [x] Do tests on Debian/Ubuntu
- [x] Improve writing

Tested personally on FreeBSD 12-CURRENT.

### CODE REVIEW

File: Monika After Story/game/script-topics.rpy [hunk 1]
```diff
@@ -4822,6 +4823,45 @@ label monika_urgent:
...
+    m "... or at least, shut it down so no data is loss"
+    return
+
+init 5 python:
+    import battery
+
+label monika_close_game_battery:
+    m "Hey [player]..."
+    m "I'm still worried about data loss."
+    m "So.. I'll just close the game for now..."
```
Comment (yukiisbored) [inline Monika After Story/game/script-topics.rpy]: I wonder if there's a way to "auto-prompt" this dialog.
I was thinking about adding a pause for 2 seconds (as if Monika is trying to find a way to close the game and recheck the battery charging status).

**Comment by yukiisbored** [comment]
Anyway, here's a short summary on how to test the battery topic.

For those who have laptops or battery powered PCs:

1. Go to the `Monika After Story/game/python-packages` directory
2. Launch a Python 2.7 REPL (`python2`)
3. First test: Get the percentage
    ```python
    import battery
    battery.get_level()
    ```
    You should see your battery percentage
4. Second Test: Check if plugged in checks work
    ```python
    import battery
    battery.is_charging()
    ```
    You should see the status if your laptop is plugged in or not, repeat this test for the other case (both plugged in and not plugged in)
5. Third Test: Check if battery is present
    ```python
    import battery
    battery.is_battery_present()
    ```
    You should see if the battery is present or not, repeat  this test for the other case (both present and not present)

For those who have desktop PCs:

Do the tests for laptops but checks if the values for `is_battery_present()` is False, `get_level()` is None, and `is_charging()` is True.

**Comment by yukiisbored** [comment]
Looks like there's a merge conflict, gonna rebase to that branch.

File: Monika After Story/game/script-topics.rpy (hunk 1 above)
Comment (ThePotatoGuy) [reply Monika After Story/game/script-topics.rpy]: add a `{nw}` to the dialogue to auto prompt. To add a pause, add `{w=#.#}`. where # is the number of seconds. For example:
```
m "I'm going to wait 2 seconds and then advance this text without you{w=2.0}{nw}"
```

**Comment by therationalpi** [comment]
I love you for rebasing.

**Comment by yukiisbored** [comment]
I'm gonna run a round of coala with PEP8Bear to make the battery module comply to PEP8 guidelines.

**Comment by supchppt** [comment]
Looks good. I was expecting someone with more python know-how to write a better script than what I could do. Other than a couple of typos, everything looks great! I'd gladly help if need be.

**Comment by ThePotatoGuy** [comment]
@supchppt , if you're using windows, mac, or linux on a laptop can you help test this?

**Comment by monikLover** [comment]
@yukiisbored I tested your code and encountered this:
image
This happened while my laptop battery is charging. This occurred when I plugged it in while the game was running and before it was running.

File: Monika After Story/game/python-packages/battery/windows.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    return percentage
+
+
+def is_charging():
+    """
+    Check if the system is charging based on the ACLineStatus.
+
+    :return: True if it's charging, false otherwise
+    """
+    return ACLineStatus == 1 or ACLineStatus == 255
```
Comment (yukiisbored) [inline Monika After Story/game/python-packages/battery/windows.py]: Whoops, a major typo here. It should be _get_ac_status().

**Comment by yukiisbored** [comment]
@monikLover I've fixed the issue, can you please try test it again?

**Comment by supchppt** [comment]
@ThePotatoGuy I run Debian Stretch on my power-hungry laptop so I think I'll give it a shot over the weekend.

File: Monika After Story/game/python-packages/battery/windows.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            ('BatteryFullLifeTime', wintypes.DWORD),
+        ]
+
+    SPS_PTR = ctypes.POINTER(SYSTEM_POWER_STATUS)
+
+    GetSystemPowerStatus = ctypes.windll.kernel32.GetSystemPowerStatus
+    GetSystemPowerStatus.argtypes = [SPS_PTR]
+    GetSystemPowerStatus.restype = wintypes.BOOL
+
+    return SYSTEM_POWER_STATUS
```
Comment (yukiisbored) [inline Monika After Story/game/python-packages/battery/windows.py]: I actually forgot to run the "GetSystemPowerStatus" function which change the value of the SYSTEM_POWER_STATUS instance in memory.

```python
status = SYSTEM_POWER_STATUS()
GetSystemPowerStatus(ctypes.pointer(status))
return status
```

**Comment by yukiisbored** [comment]
Confirmed works on Windows 10

a

I can't test battery present since this laptop have the battery planted

**Comment by ThePotatoGuy** [comment]
Thanks @supchppt . 

Also thanks for your continued work @yukiisbored . I don't have a laptop that can run renpy so I've been unable to test anything myself. I've requested testing from the rest of the dev team but we're all pretty busy with different things atm.

**Comment by supchppt** [comment]
I am getting this. Might want to check the way packages are imported.
screenshot from 2018-01-13 03-53-00

**Comment by yukiisbored** [comment]
Hmm, that's odd. It should not execute that and I already add a try/except in case wintypes doesn't exist (it's a Windows only package).




--


Muhammad Kaisar Arkhan







On Sat, Jan 13, 2018 at 3:29 AM +0700, "supchppt" <notifications@github.com> wrote:












I am getting this. Might want to check the way packages are imported.

**Comment by yukiisbored** [comment]
Ah, nvm, I just noticed it's caused by moving the class to the module scope

**Comment by yukiisbored** [comment]
It should be fixed now, @supchppt.

**Comment by supchppt** [comment]
@yukiisbored I don't know why this happened, but this is some JavaScript-grade nonsense.
[traceback.txt](https://github.com/Backdash/MonikaModDev/files/1628339/traceback.txt)

**Comment by yukiisbored** [comment]
It doesn't look like my patch caused that. Probably some garbage got into the event pool.




Try moving your saves, maybe? (`mv ~/.renpy ~/old_renpy_saves`)




--


Muhammad Kaisar Arkhan







On Sat, Jan 13, 2018 at 7:52 PM +0700, "supchppt" <notifications@github.com> wrote:












@yukiisbored I don't know why this happened, but this is some JavaScript-grade nonsense.

traceback.txt

**Comment by supchppt** [comment]
I was too eager to test the PR out that I did not inspect the actual scripts. Now that I have the time to do so, I found a couple of errors.

0. In `get_level()`, you are trying to read from a directory. The actual file to be read is `$bat/capacity`.

1. In `is_charging()`, the path does not exist, at least in by my case. The path you are looking for should be `$bat/status`. Consequently, Monika's dialogue will never be reached because it defaults to `True`.

Replace `$bat` with the value returned by `_get_battery()`. Here's a screenshot to demonstrate.

Thanks for the good work!
screenshot from 2018-01-14 21-04-48

**Comment by yukiisbored** [comment]
Thank you @supchppt, I just fixed the bugs you've reported. Can you test it again, please? <3

**Comment by supchppt** [comment]
There is a typo on line 36 of linux.py. Could you fix that? It works with the typo fixed.

**Comment by supchppt** [comment]
@yukiisbored, could you fix the issue described? Thank you~

**Comment by yukiisbored** [comment]
@supchppt I did. I fixup'd my commit almost a week ago resolving your issue.

**Comment by supchppt** [comment]
Sorry, I didn't notice. You can check Linux off the todo list now.

**Comment by kaido1224** [comment]
Tested on a PC laptop and everything seems to be working the way it should be.

**Comment by monikLover** [comment]
Dialogue has been fleshed-out and tests on unplugged and desktop windows have been conducted. Just waiting on @aldoram5 to test it out on macOS.

**Comment by IsItBadToMakeAGitHubAccountJustForThis** [comment]
Just an "educated" guess, but won't `monika_close_game_battery` repeat if the battery is plugged/unplugged?

**Comment by yukiisbored** [comment]
@IsItBadToMakeAGitHubAccountJustForThis What do you mean by plugged/unplugged?

**Comment by IsItBadToMakeAGitHubAccountJustForThis** [comment]
I mean plugged/unplugged from the charger.
For example, if somebody has a faulty wire and it sometimes doesn't provide a consistent connection.
But now that I think about it. How does it behave in systems with multiple batteries? (Like Thinkpads.)

**Comment by IsItBadToMakeAGitHubAccountJustForThis** [comment]
Still some problems on Linux according to #1285.

**Pull request merged**